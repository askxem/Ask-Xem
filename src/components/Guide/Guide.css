.guideContainer {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  /* display: grid;
  grid-template-rows: 1fr 1fr;
  grid-template-columns: 1fr 1fr; */
  width: 90vw;
  margin-right: 5%;
}
.guideText {
  height: auto;
	padding: 30px;
	margin: 40px auto;
  position: relative;
	border-radius: 200px;
	text-align: center;
	border: 1px solid #333;
  background-color: white;
}

.guideText:before,
.guideText:after {
	content: ' ';
	position: absolute;
	left: 10px;
	bottom: -30px;
	width: 40px;
	height: 40px;
	background-color: white;
	border: 1px solid #333;
	border-radius: 28px;
}

.guideText:after {
	width: 20px;
	height: 20px;
	left: 5px;
	bottom: -40px;
	border-radius: 18px;
  background-color: white;
}

.guideImg{
  width: 150px;
  position: absolute;
  padding: 10px;
  border-radius: 10%;
  position: relative;
  padding: 0;
  margin: 0;
  justify-content: center;
} 

