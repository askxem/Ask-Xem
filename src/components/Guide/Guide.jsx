import PropTypes from "prop-types";
import { useState } from "react";

export default function Guide({text}) {
  //LATER:
    // pull users guide from user context
  const [userGuide, setUserGuide] = useState('lion');
  

  return (
    <article>
      <div aria-label='guide text box'>
        <pre>
        {text}
        </pre>
      </div>
      <img src={require(`../../assets/Icons/${userGuide}.png`)} alt={userGuide} aria-label='user guide'/>
    </article>
  )
}

Guide.propTypes = {
  text: PropTypes.string,
}