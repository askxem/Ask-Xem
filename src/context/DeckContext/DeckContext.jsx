import { createContext, useState, useContext } from "react";

const DeckContext = createContext()

const DeckProvider = ({ children }) => {
    const [genSeen, setGenSeen] = useState(JSON.parse(localStorage.getItem('genSeen')) || [])
    const [pronSeen, setPronSeen] = useState(JSON.parse(localStorage.getItem('pronSeen')) || [])

    const seen = (id, category) => {

        if(category == 'gender') {
            let genSet = new Set(genSeen)
            genSet.add(id)
            const arr = Array.from(genSet)
            setGenSeen(arr)
            localStorage.setItem('genSEEN', JSON.stringify(arr))
        }
        if(category == 'pronoun') {
            let pronSet = new Set(pronSeen)
            pronSet.add(id)
            const arr = Array.from(pronSet)
            setPronSeen(arr)
            localStorage.setItem('pronSEEN', JSON.stringify(arr))
        }
    }

    return(
        <DeckContext.Provider value={{ genSeen, pronSeen, seen }}>{children}</DeckContext.Provider>
    )

}

const useDeck = () => {
    const context = useContext(DeckContext)
    if (context === undefined) {
        throw new Error('useDeck not used inside provider')
    }
    return context
}

export { DeckProvider, useDeck }