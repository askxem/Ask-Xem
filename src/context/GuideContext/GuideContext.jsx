import { createContext, useState, useContext } from "react";

const GuideContext = createContext();

function GuideProvider({children}) {

  const hasSelectedGuide = getGuideLocal();

  const [guide, setGuide] = useState(hasSelectedGuide ? hasSelectedGuide : '');

    return (
        <GuideContext.Provider value={{guide, setGuide}} >
            {children}
        </GuideContext.Provider>
    )

function setGuideGlobal(guide) {
  setGuideLocal(guide);
  setGuide(guide);
}

function getGuideLocal() {
  const stringyGuide = localStorage.getItem('GUIDE');
  return JSON.parse(stringyGuide);
}

function setGuideLocal(guide) {
  const stringyGuide = JSON.stringify(guide);
  localStorage.setItem('GUIDE', stringyGuide)
}

function useGuide(){
    const context = useContext(GuideContext);
    if (context === undefined) throw new Error('Guide context not accessible outside of the guide provider.')
    return context;
}
}

export { GuideProvider, useGuide, setGuideGlobal }