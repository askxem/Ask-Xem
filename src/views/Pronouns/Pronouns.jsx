import { getCardsbyCategory } from "../../services/cards"
import { useState, useEffect } from "react"
import CardList from "../../components/Cards/CardList"

export default function Pronouns() {
  const [deck, setDeck] = useState('')
  const [loading, setLoading] =useState(true)

  useEffect(() => {
    const fetchDeck = async () => {
      try {
        setLoading(true)
        const response = await getCardsbyCategory('pronoun')
        setDeck(response)
        setLoading(false)
      } catch (error) {
       console.log(error.message) 
      }
    }
    fetchDeck()
  }, [])

  return (
    <main>
      {loading && <p>Loading...</p>}
      <CardList cards={deck}/>
    </main>
  )
}
