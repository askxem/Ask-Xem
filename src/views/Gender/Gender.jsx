import { useState, useEffect } from "react"
import { getCardsByCategory } from "../../services/cards"
import CardList from "../../components/Cards/CardList"
import DeckComplete from "../../components/DeckComplete/DeckComplete.jsx"
import { useDeck } from "../../context/DeckContext/DeckContext"
import renderRainbow from "../../utils/rainbow"
import Guide from "../../components/Guide/Guide.jsx"
import Loader from "../../components/Loading/Loader.jsx"



export default function Gender() {
  const [deck, setDeck] = useState('')
  const [loading, setLoading] =useState(true)
  const [rainbow, setRainbow] = useState('')
  const [showModal, setShowModal] = useState(true)
  
  const { genSeen } = useDeck()

  const guideText = 'This is the Gender Deck - click on a card to find out more!'

  const handleClick = () => {
    setShowModal(false)
  }

  useEffect(() => {
    const showModal = () => {
      if (genSeen.length === 8) {
        setShowModal(true)
      }
    }
    showModal()
  },[genSeen])

  useEffect(() => {
    const fetchDeck = async () => {
      setLoading(true)

      try {
        const response = await getCardsByCategory('gender')
        setDeck(response)
        setLoading(false)
      } catch (error) {
       console.log(error.message) 
      }
    }
    fetchDeck()
  }, [])

  useEffect(() => {
    const showRainbow = () => {
      const rainbow = renderRainbow(genSeen.length)
      setRainbow(rainbow)
    }
    showRainbow()
  }, [genSeen])

  return (
    <main>
      {loading && <Loader />}
      <CardList cards={deck} rainbow={rainbow} />
      <Guide text={guideText} />
      { showModal && <DeckComplete handleClick={handleClick}/>}
    </main>
  )
}