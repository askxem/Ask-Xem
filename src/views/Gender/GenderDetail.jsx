import { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import CardBack from '../../components/Cards/CardBack'
import { getCard } from '../../services/cards'
import { useDeck } from '../../context/DeckContext/DeckContext'


export default function GenderDetail() {
  const { id } = useParams();
  const [loading, setLoading] = useState(true)
  const [card, setCard] = useState(null)
  const { seen } = useDeck()


  useEffect(() => {
    const fetchCard = async () => {
      setLoading(true)

      try {
        const [response] = await getCard(id)
        setCard(response)
        setLoading(false)
        seen(response.id, response.category)
      } catch (error) {
       console.log(error.message) 
      }
    }
    fetchCard()
  }, [])


  return (
    <main>
      {loading && <p>Loading...</p>}
      {card && <CardBack card={card}/>}
    </main>
  )
}
