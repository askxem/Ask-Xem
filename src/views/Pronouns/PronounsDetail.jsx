import { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import CardBack from '../../components/Cards/CardBack'
import Guide from '../../components/Guide/Guide.jsx'
import retrieveGuideText from '../../utils/retrieveGuideText/retrieveGuideText.js'
import { getCard } from '../../services/cards'
import { getFavs } from '../../services/favorites'
import { useDeck } from '../../context/DeckContext/DeckContext'
import { useAuth } from '../../context/AuthContext'
import Loader from '../../components/Loading/Loader.jsx'

export default function PronounsDetail() {
  const { user } = useAuth()
  const { id } = useParams();
  const [loading, setLoading] = useState(true)
  const [card, setCard] = useState(null)
  const [guideText, setGuideText] = useState('')
  const { seen } = useDeck()
  const [favStatus, setFavStatus] = useState(false)

  useEffect(() => {
    const fetchFavs = async () => {
      if (user.id) {
        try {
          //get the card ids in the favs table for this user
          const response = await getFavs(user.id)
          //extract an array of just the card ids from the response         
          const favArr = response.map((item) => item.card_id)
          //does the extracted array include the current card?
          const faved = favArr.includes(Number(id))
          //set favStatus to pass to card component so appropriate heart is displayed
          setFavStatus(faved)
        } catch (error) {
          console.log(error.message)
        }
      }
    }
    fetchFavs()
  }, [])

  useEffect(() => {
    const fetchCard = async () => {
      setLoading(true)

      try {
        const [response] = await getCard(id)
        setCard(response)
        setLoading(false)
        seen(response.id, response.category)
        const newGuideText = await retrieveGuideText(response.category, response.animal, response.title)
        setGuideText(newGuideText)
      } catch (error) {
        console.log(error.message)
      }
    }
    fetchCard()
  }, [])


  return (
    <main>
      {loading && <Loader />}
      {card && <CardBack card={card} favStatus={favStatus} />}
      {guideText && <Guide text={guideText} />}
    </main>
  )
}
