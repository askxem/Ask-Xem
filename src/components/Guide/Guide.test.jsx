import { render, screen, waitFor } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { GuideProvider } from "../../context/GuideContext/GuideContext.jsx"
import { DeckProvider } from "../../context/DeckContext/DeckContext.jsx"
import Gender from "../../views/Gender/Gender.jsx"
import Guide from "./Guide.jsx"
import Pronouns from "../../views/Pronouns/Pronouns.jsx"

it('renders guide component', () => {
    const { container } = render(
        <GuideProvider>
            <Guide text='test-text'/>
        </GuideProvider>
    );
    expect(container).toMatchSnapshot();
});

it('should render the Gender guide text message', async() => {
  render(
      <GuideProvider>
        <DeckProvider>
            <MemoryRouter initialEntries={['/gender']}>
                <Gender />
            </MemoryRouter>
          </DeckProvider>
      </GuideProvider>
  );

screen.getByLabelText(/loading./i)

const guideText = "This is the Gender Deck - click on a card to find out more! Visit all of my friends for a colorful surprise :)"

await waitFor(() => screen.findByText(guideText));
}); 

it('when on the Pronouns page - should render the Pronouns guide text message', async() => {
  render(
    <MemoryRouter initialEntries={['/pronouns']}>
      <DeckProvider>
        <GuideProvider>
            <Pronouns />
        </GuideProvider>
      </DeckProvider>
    </MemoryRouter>
  )

  screen.getByLabelText(/loading./i)

  const guideText = "This is the Pronoun Deck - click on a card to find out more! Visit all of my friends for a colorful surprise :)"

  await waitFor(() => screen.findByText(guideText));
});