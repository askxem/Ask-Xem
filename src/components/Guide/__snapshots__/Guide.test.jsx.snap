// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders guide component 1`] = `
<div>
  <article
    class="guideContainer"
  >
    <div
      aria-label="guide text box"
      class="guideText"
    >
      <p>
        test-text
      </p>
    </div>
    <img
      alt="lion"
      aria-label="user guide"
      class="guideImg"
      src="/icons/lion.png"
    />
  </article>
</div>
`;
